modules/navigation/2d/godot_navigation_server_2d.cpp

```diff
@@ -456,16 +452,48 @@ Vector<Vector2> GodotNavigationServer2D::obstacle_get_vertices(RID p_obstacle) c
 	return vector_v3_to_v2(NavigationServer3D::get_singleton()->obstacle_get_vertices(p_obstacle));
 }
 
-void GodotNavigationServer2D::query_path(const Ref<NavigationPathQueryParameters2D> &p_query_parameters, Ref<NavigationPathQueryResult2D> p_query_result) const {
+void GodotNavigationServer2D::query_path(const Ref<NavigationPathQueryParameters2D> &p_query_parameters, Ref<NavigationPathQueryResult2D> p_query_result, const Callable &p_callback) {
 	ERR_FAIL_COND(!p_query_parameters.is_valid());
 	ERR_FAIL_COND(!p_query_result.is_valid());
 
-	const NavigationUtilities::PathQueryResult _query_result = NavigationServer3D::get_singleton()->_query_path(p_query_parameters->get_parameters());
+	Ref<NavigationPathQueryParameters3D> query_parameters;
+	query_parameters.instantiate();
+
+	query_parameters->set_map(p_query_parameters->get_map());
+	query_parameters->set_start_position(v2_to_v3(p_query_parameters->get_start_position()));
+	query_parameters->set_target_position(v2_to_v3(p_query_parameters->get_target_position()));
+	query_parameters->set_navigation_layers(p_query_parameters->get_navigation_layers());
+	query_parameters->set_pathfinding_algorithm(NavigationPathQueryParameters3D::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR);
+
+	switch (p_query_parameters->get_path_postprocessing()) {
+		case NavigationPathQueryParameters2D::PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL: {
+			query_parameters->set_path_postprocessing(NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL);
+		} break;
+		case NavigationPathQueryParameters2D::PathPostProcessing::PATH_POSTPROCESSING_EDGECENTERED: {
+			query_parameters->set_path_postprocessing(NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_EDGECENTERED);
+		} break;
+		case NavigationPathQueryParameters2D::PathPostProcessing::PATH_POSTPROCESSING_NONE: {
+			query_parameters->set_path_postprocessing(NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_NONE);
+		} break;
+		default: {
+			WARN_PRINT("No match for used PathPostProcessing - fallback to default");
+			query_parameters->set_path_postprocessing(NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL);
+		} break;
+	}
+
+	query_parameters->set_metadata_flags((int64_t)p_query_parameters->get_metadata_flags());
+	query_parameters->set_simplify_path(p_query_parameters->get_simplify_path());
+	query_parameters->set_simplify_epsilon(p_query_parameters->get_simplify_epsilon());
+
+	Ref<NavigationPathQueryResult3D> query_result;
+	query_result.instantiate();
+
+	NavigationServer3D::get_singleton()->query_path(query_parameters, query_result, p_callback);
```

Comment:
Same as with the Parameters, things like this that are exposed to script users and the binding system need to be RefCounted. To avoid new object creation all the time we just update the objects in the query instead of returning new things. There are older discussions e.g. physics ray API, that returns stuff which is really bad for performance, especially with languages like C# that struggle with their garbage collector.